src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -320,24 +336,46 @@ private void EventLoop()
                     // The native shim is responsible for ensuring this condition.
                     Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
 
+                    bool enqueuedEvent = false;
                     for (int i = 0; i < numEvents; i++)
                     {
-                        IntPtr handle = _buffer[i].Data;
+                        IntPtr handle = buffer[i].Data;
                         if (handle == ShutdownHandle)
                         {
                             shutdown = true;
                         }
                         else
                         {
                             Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
-                            _handleToContextMap.TryGetValue(handle, out SocketAsyncContext? context);
+                            handleToContextMap.TryGetValue(handle, out SocketAsyncContext? context);
                             if (context != null)
                             {
-                                context.HandleEvents(_buffer[i].Events);
+                                Interop.Sys.SocketEvents events = buffer[i].Events;
+                                events = context.HandleSyncEventsSpeculatively(events);
+                                if (events != Interop.Sys.SocketEvents.None)
+                                {
+                                    var ev = new SocketIOEvent(context, events);
+                                    eventQueue.Enqueue(ev);
+                                    enqueuedEvent = true;
+
+                                    // This is necessary when the JIT generates unoptimized code (debug builds, live debugging,
+                                    // quick JIT, etc.) to ensure that the context does not remain referenced by this method, as
+                                    // such code may keep the stack location live for longer than necessary
+                                    ev = default;
```

Comment:
Would not using a local variable avoid the problem?

```cs
eventQueue.Enqueue(new SocketIOEvent(context, events));
```

Reply:
That's what I was doing initially, it still created an implicit local that was kept alive